File: ui/src/lib/ipv6names.ts
import { createHash } from 'crypto';
import ipv6NamesData from '../../../tui/ipv6names.json';

interface IPv6NamesData {
  adjectives: Record<string, string[]>;
  nouns: Record<string, string[]>;
  celestialTypes: Record<string, string[]>;
  subnetMappings: Record<string, number>;
  levelNames: string[];
}

const namesData = ipv6NamesData as IPv6NamesData;

const adjectives: Record<number, string[]> = {};
const nouns: Record<number, string[]> = {};
const celestialTypes: Record<number, string[]> = {};

for (const [strKey, value] of Object.entries(namesData.adjectives)) {
  adjectives[parseInt(strKey)] = value;
}

for (const [strKey, value] of Object.entries(namesData.nouns)) {
  nouns[parseInt(strKey)] = value;
}

for (const [strKey, value] of Object.entries(namesData.celestialTypes)) {
  celestialTypes[parseInt(strKey)] = value;
}

// Subnet size mappings from JSON data
export const subnetMappings: Record<number, number> = {};
for (const [strKey, value] of Object.entries(namesData.subnetMappings)) {
  subnetMappings[parseInt(strKey)] = value;
}

export const levelNames = namesData.levelNames;

// Helper function to parse IPv6 address and create truncated version
function truncateIPv6(addr: string, bits: number): Buffer {
  // Parse IPv6 address to bytes
  const parts = addr.split(':');
  const bytes = Buffer.alloc(16);
  
  // Handle compressed notation
  let expandedParts: string[] = [];
  const compressedIndex = parts.indexOf('');
  
  if (compressedIndex !== -1) {
    const beforeCompressed = parts.slice(0, compressedIndex);
    const afterCompressed = parts.slice(compressedIndex + 1).filter(p => p !== '');
    const missingParts = 8 - beforeCompressed.length - afterCompressed.length;
    
    expandedParts = [
      ...beforeCompressed,
      ...Array(missingParts).fill('0000'),
      ...afterCompressed
    ];
  } else {
    expandedParts = parts;
  }
  
  // Fill in any missing parts with zeros
  while (expandedParts.length < 8) {
    expandedParts.push('0000');
  }
  
  // Convert to bytes
  for (let i = 0; i < 8; i++) {
    const part = parseInt(expandedParts[i] || '0000', 16);
    bytes.writeUInt16BE(part, i * 2);
  }
  
  // Apply subnet mask
  const maskBytes = Math.floor(bits / 8);
  const maskBits = bits % 8;
  
  // Zero out bytes beyond the mask
  for (let i = maskBytes + (maskBits > 0 ? 1 : 0); i < 16; i++) {
    bytes[i] = 0;
  }
  
  // Apply bit mask to the last partial byte
  if (maskBits > 0) {
    const mask = (0xFF << (8 - maskBits)) & 0xFF;
    bytes[maskBytes] &= mask;
  }
  
  return bytes;
}

// Generate a unique name for an IPv6 address at a given subnet size
export function generateName(addr: string, subnetSize: number): string {
  if (!celestialTypes[subnetSize]) {
    throw new Error(`Invalid subnet size: ${subnetSize}`);
  }
  
  // Create a hash of the truncated address based on subnet size
  const truncatedAddr = truncateIPv6(addr, subnetSize);
  const subnetSizeBuffer = Buffer.alloc(4);
  subnetSizeBuffer.writeUInt32BE(subnetSize, 0);
  
  const hash = createHash('sha256')
    .update(truncatedAddr)
    .update(subnetSizeBuffer)
    .digest();
  
  // Use different parts of the hash for different components of the name
  const adjIndex = hash.readUInt32BE(0) % adjectives[subnetSize].length;
  const nounIndex = hash.readUInt32BE(4) % nouns[subnetSize].length;
  const celestialIndex = hash.readUInt32BE(8) % celestialTypes[subnetSize].length;
  
  // Generate a numeric suffix using another part of the hash
  const suffix = hash.readUInt16BE(12) % 1000;
  
  return `${adjectives[subnetSize][adjIndex]} ${nouns[subnetSize][nounIndex]} ${celestialTypes[subnetSize][celestialIndex]}-${suffix}`;
}

// Generate the full hierarchy for an IPv6 address
export function getHierarchy(addr: string): string[] {
  const sizes = [16, 32, 48, 64, 80, 96, 112, 128];
  return sizes.map(size => generateName(addr, size));
}

// Create a full IPv6 address for a given index and level
export function makeIPv6Full(index: number, prefix: string, level: number): { addr: string; subnet: number } {
  const hex = index.toString(16).padStart(4, '0');
  const numSubBlocks = 8 - (level + 1);
  const zeroBlocks = ':0000'.repeat(numSubBlocks);
  
  const full = prefix ? `${prefix}${hex}${zeroBlocks}` : `${hex}${zeroBlocks}`;
  return {
    addr: full,
    subnet: subnetMappings[level]
  };
}

export type SubnetLevel = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;

export interface SubnetInfo {
  name: string;
  addr: string;
  subnet: number;
  level: SubnetLevel;
  levelName: string;
}

File: tui/ipv6names.go
package main

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"net"
	"strings"
	_ "embed"
)

//go:embed ipv6names.json
var ipv6NamesData []byte

type IPv6Names struct {
	Adjectives     map[string][]string `json:"adjectives"`
	Nouns          map[string][]string `json:"nouns"`
	CelestialTypes map[string][]string `json:"celestialTypes"`
	SubnetMappings map[string]int      `json:"subnetMappings"`
	LevelNames     []string            `json:"levelNames"`
}

var (
	adjectives     map[int][]string
	nouns          map[int][]string
	celestialTypes map[int][]string
)

func init() {
	var namesData IPv6Names
	if err := json.Unmarshal(ipv6NamesData, &namesData); err != nil {
		panic(fmt.Sprintf("Failed to parse embedded IPv6 names data: %v", err))
	}

	adjectives = make(map[int][]string)
	nouns = make(map[int][]string)
	celestialTypes = make(map[int][]string)

	for strKey, value := range namesData.Adjectives {
		var key int
		fmt.Sscanf(strKey, "%d", &key)
		adjectives[key] = value
	}

	for strKey, value := range namesData.Nouns {
		var key int
		fmt.Sscanf(strKey, "%d", &key)
		nouns[key] = value
	}

	for strKey, value := range namesData.CelestialTypes {
		var key int
		fmt.Sscanf(strKey, "%d", &key)
		celestialTypes[key] = value
	}
}

// GenerateName creates a unique name for an IPv6 address at a given subnet size
func GenerateName(saddr string, subnetSize int) (string, error) {
	addr := net.ParseIP(saddr)
	if addr == nil {
		return "", fmt.Errorf("invalid IPv6 address")
	}
	if addr.To16() == nil {
		return "", fmt.Errorf("invalid IPv6 address")
	}
	if _, ok := celestialTypes[subnetSize]; !ok {
		return "", fmt.Errorf("invalid subnet size: %d", subnetSize)
	}

	// Create a hash of the truncated address based on subnet size
	truncatedAddr := truncateIPv6(addr, subnetSize)
	subnetSizeBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(subnetSizeBytes, uint32(subnetSize))
	hash := sha256.Sum256(append(truncatedAddr, subnetSizeBytes...))

	// Use different parts of the hash for different components of the name
	adjIndex := int(binary.BigEndian.Uint32(hash[0:4])) % len(adjectives[subnetSize])
	nounIndex := int(binary.BigEndian.Uint32(hash[4:8])) % len(nouns[subnetSize])
	celestialIndex := int(binary.BigEndian.Uint32(hash[8:12])) % len(celestialTypes[subnetSize])

	// Generate a numeric suffix using another part of the hash
	suffix := binary.BigEndian.Uint16(hash[12:14]) % 1000

	return fmt.Sprintf("%s %s %s-%d",
		adjectives[subnetSize][adjIndex],
		nouns[subnetSize][nounIndex],
		celestialTypes[subnetSize][celestialIndex],
		suffix,
	), nil
}

// truncateIPv6 masks an IPv6 address to the specified subnet size
func truncateIPv6(addr net.IP, bits int) []byte {
	mask := net.CIDRMask(bits, 128)
	return addr.Mask(mask)
}

// GetHierarchy generates names for all parent categories of an IPv6 address
func GetHierarchy(addr net.IP) ([]string, error) {
	sizes := []int{16, 32, 48, 64, 80, 96, 112, 128}
	var names []string

	for _, size := range sizes {
		name, err := GenerateName(addr.String(), size)
		if err != nil {
			return nil, err
		}
		names = append(names, name)
	}

	return names, nil
}

// FormatHierarchy returns a formatted string of the complete hierarchy
func FormatHierarchy(addr net.IP) (string, error) {
	names, err := GetHierarchy(addr)
	if err != nil {
		return "", err
	}

	var builder strings.Builder
	for i, name := range names {
		builder.WriteString(strings.Repeat("  ", i))
		builder.WriteString(name)
		builder.WriteString("\n")
	}

	return builder.String(), nil
}
